core/domain/expression_parser.py

```diff
@@ -0,0 +1,541 @@
+            raise Exception('Invalid bracket pairing.')
+
+        # Replacing all parens with '(' or ')' for simplicity.
+        for i in python_utils.RANGE(len(self.token_list)):
+            token = self.token_list[i]
+            if token.text in ['[', '{']:
+                self.token_list[i] = Token('(')
+            if token.text in ['}', ']']:
+                self.token_list[i] = Token(')')
+
+    def parse(self):
+        """A wrapper around the parse_expr method. This method checks if all
+        tokens have been consumed by the parse_expr method.
+
+        Returns:
+            Node. Root node of the generated parse tree.
+
+        Raises:
+            Exception: Invalid syntax.
+        """
+
+        parsed_expr = self._parse_expr()
+        if self.next_token_index < len(self.token_list):
+            raise Exception('Invalid syntax.')
+        return parsed_expr
+
+    def _parse_expr(self):
+        """Function representing the following production rule of the grammar:
+        <expr> ::= <mul_expr> (('+' | '-') <mul_expr>)*
+
+        Returns:
+            Node. Root node of the generated parse tree.
+        """
+        parsed_expr = self.parse_mul_expr()
+        operator_token = self.is_next_token(['+', '-'])
+        while operator_token:
+            parsed_right = self.parse_mul_expr()
+            if operator_token.text == '+':
+                parsed_expr = AddOperator(parsed_expr, parsed_right)
+            else:
+                parsed_expr = SubtractOperator(parsed_expr, parsed_right)
+            operator_token = self.is_next_token(['+', '-'])
+        return parsed_expr
+
+    def parse_mul_expr(self):
+        """Function representing the following production rule of the grammar:
+        <mul_expr> ::= <pow_expr> (('*' | '/') <pow_expr>)*
+
+        Returns:
+            Node. Root node of the generated parse tree.
+        """
+
+        parsed_expr = self.parse_pow_expr()
+        operator_token = self.is_next_token(['*', '/'])
+        while operator_token:
+            parsed_right = self.parse_pow_expr()
+            if operator_token.text == '*':
+                parsed_expr = MultiplyOperator(parsed_expr, parsed_right)
+            else:
+                parsed_expr = DivideOperator(parsed_expr, parsed_right)
+            operator_token = self.is_next_token(['*', '/'])
+        return parsed_expr
+
+    def parse_pow_expr(self):
+        """Function representing the following production rule of the grammar:
+        <pow_expr> ::= '-' <pow_expr> | '+' <pow_expr> |
+        <unit> ('^' <pow_expr>)?
+
+        Returns:
+            Node. Root node of the generated parse tree.
+        """
+
+        # Eliminate any leading unary '+' or '-' operators, because they
+        # are syntactically irrelevant.
+        while self.is_next_token(['+', '-']):
+            pass
+
+        parsed_expr = self.parse_unit()
+        operator_token = self.is_next_token(['^'])
```

Comment:
`is_next_token` means it is just checking whether the next is this operator. Either use get_next_token here or combine both these functions.

Reply:
I've renamed the function to avoid confusion.